// panasonic-comfort-cloud-api/dist/ComfortCloud.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComfortCloud = void 0;
const https = __importStar(require("https"));
const http_1 = __importDefault(require("http"));
const url = __importStar(require("url"));
const HttpMethod_1 = require("./models/HttpMethod");
const enums_1 = require("./models/enums");
const LoginRequest_1 = require("./models/LoginRequest");
class ComfortCloud {
    get token() {
        return this._accessToken;
    }
    set token(value) {
        this._accessToken = value;
    }
    get clientId() {
        return this._clientId;
    }
    set clientId(value) {
        this._clientId = value;
    }
    get ccAppVersion() {
        return this._ccAppVersion;
    }
    set ccAppVersion(value) {
        this._ccAppVersion = value;
    }
    constructor(username, password) {
        this._config = {
            username: "",
            password: "",
            cc_app_url: "https://itunes.apple.com/lookup?id=1348640525",
            base_url: "https://accsmart.panasonic.com",
            login_url: "/auth/login",
            group_url: "/device/group",
            device_url: "/deviceStatus/{guid}",
            device_now_url: "/deviceStatus/now/{guid}",
            device_control_url: "/deviceStatus/control",
            device_history_url: "/deviceHistoryData",
        };
        this._accessToken = "";
        this._clientId = "";
        this._config.username = username;
        this._config.password = password;
        this._ccAppVersion = "1.17.0"; // This value will be replaced by value returned from Apple App Store.
    }
    /**
     * Returns Panasonic Comfort Cloud app version
     * @returns Version number as String
     */
    getCcAppVersion() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const uri = url.parse(`${this._config.cc_app_url}`, true);
                const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Get, uri);
                const response = yield this.request(options);
                const version = response.results[0].version;
                return version;
            }
            catch (error) {
                console.error(error);
            }
            return this._ccAppVersion;
        });
    }
    /**
     * Login to Panasonic Comfort Cloud
     * @param username Login username
     * @param password Login password
     * @returns LoginResponse containing access token.
     */
    login(username = this._config.username, password = this._config.password) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!username || !password)
                throw new Error("Username and password must contain a value.");
            this._ccAppVersion = yield this.getCcAppVersion();
            const data = new LoginRequest_1.LoginRequest(username, password);
            const uri = url.parse(`${this._config.base_url}${this._config.login_url}`, true);
            const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Post, uri);
            const result = yield this.request(options, JSON.stringify(data));
            if (result === null || result === void 0 ? void 0 : result.uToken) {
                this._accessToken = result.uToken;
                this._clientId = result.clientId;
                return result;
            }
            return undefined;
        });
    }
    /**
     * Get groups.
     * @returns A list of groups containing list of devices.
     */
    getGroups() {
        return __awaiter(this, void 0, void 0, function* () {
            const uri = url.parse(`${this._config.base_url}${this._config.group_url}`, true);
            const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Get, uri);
            const result = yield this.request(options);
            if ((result === null || result === void 0 ? void 0 : result.groupCount) > 0) {
                const data = result;
                return data.groupList;
            }
            return [];
        });
    }
    /**
     * Returns device with the provided Device ID
     * @param deviceId Device ID to use. Aka deviceGuid
     * @returns Device based on deviceId
     */
    getDevice(deviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const uri = url.parse(`${this._config.base_url}${this._config.device_url.replace("{guid}", deviceId)}`, true);
            const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Get, uri);
            const result = yield this.request(options);
            result.deviceGuid = deviceId;
            return result;
        });
    }
    /**
     * Returns device with the provided Device ID
     * @param deviceId Device ID to use. Aka deviceGuid
     * @returns Device based on deviceId
     */
    getDeviceNow(deviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const uri = url.parse(`${this._config.base_url}${this._config.device_now_url.replace("{guid}", deviceId)}`, true);
            const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Get, uri);
            const result = yield this.request(options);
            result.deviceGuid = deviceId;
            return result;
        });
    }
    /**
     * Set parameters on device. Parameters can contain one or more properties.
     * @param deviceId Device ID to use. Aka deviceGuid
     * @param parameters Parameters to set
     * @returns
     */
    setParameters(deviceId, parameters) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const uri = url.parse(`${this._config.base_url}${this._config.device_control_url}`, true);
                const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Post, uri);
                const requestBody = { deviceGuid: deviceId, parameters: this.getParameters(parameters) };
                const result = yield this.request(options, JSON.stringify(requestBody));
                const response = { status: -1, statusText: "Unknown response" };
                if (result) {
                    response.status = result.result;
                    response.statusText = "OK";
                    return response;
                }
                return response;
            }
            catch (error) {
                if (error.statusCode === 401 || error.statusCode === 403)
                    throw error;
                error.statusText = "Invalid parameter. Please check the input and use the Device's boolean values to check valid capabilities.";
                throw error;
            }
        });
    }
    /**
     * Set parameters from device. Manipulate device parameters before calling.
     * @param device Device ID to be used
     * @returns true if recuest succeeds
     */
    setDevice(device) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const parameters = this.getDeviceParameters(device);
                const res = yield this.setParameters(device.deviceGuid, parameters);
                return res;
            }
            catch (error) {
                const res = { status: -1 };
                if (error && error.code && error.message) {
                    res.error = error;
                    res.status = error.code;
                }
                return res;
            }
        });
    }
    getDeviceHistory(deviceId, dataMode = 0, date, timezone = "+00:00") {
        return __awaiter(this, void 0, void 0, function* () {
            const uri = url.parse(`${this._config.base_url}${this._config.device_history_url}`, true);
            const options = this.getRequestOptions(HttpMethod_1.HttpMethod.Post, uri);
            const requestBody = { deviceGuid: deviceId, dataMode: dataMode, date: date, osTimezone: timezone };
            const result = yield this.request(options, JSON.stringify(requestBody));
            return result;
        });
    }
    getDeviceParameters(device) {
        const parameters = {};
        if (!device.parameters)
            return parameters;
        if (device.parameters.operate !== undefined)
            parameters.operate = device.parameters.operate;
        if (device.parameters.temperatureSet !== undefined)
            parameters.temperatureSet = device.parameters.temperatureSet;
        if (device.parameters.fanAutoMode !== undefined)
            parameters.fanAutoMode = device.parameters.fanAutoMode;
        if (device.parameters.airDirection !== undefined)
            parameters.airDirection = device.parameters.airDirection;
        if (device.parameters.airSwingLR !== undefined)
            parameters.airSwingLR = device.parameters.airSwingLR;
        if (device.parameters.airSwingUD !== undefined)
            parameters.airSwingUD = device.parameters.airSwingUD;
        if (device.parameters.fanSpeed !== undefined)
            parameters.fanSpeed = device.parameters.fanSpeed;
        if (device.parameters.ecoFunctionData !== undefined)
            parameters.ecoFunctionData = device.parameters.ecoFunctionData;
        if (device.parameters.ecoMode !== undefined)
            parameters.ecoMode = device.parameters.ecoMode;
        if (device.nanoeStandAlone && device.parameters.actualNanoe !== undefined)
            parameters.actualNanoe = device.parameters.actualNanoe;
        if (device.nanoe && device.parameters.nanoe !== undefined)
            parameters.nanoe = device.parameters.nanoe;
        if ((device.autoMode && device.parameters.operationMode === enums_1.OperationMode.Auto) ||
            (device.coolMode && device.parameters.operationMode === enums_1.OperationMode.Cool) ||
            (device.dryMode && device.parameters.operationMode === enums_1.OperationMode.Dry) ||
            (device.heatMode && device.parameters.operationMode === enums_1.OperationMode.Heat) ||
            (device.fanMode && device.parameters.operationMode === enums_1.OperationMode.Fan))
            parameters.operationMode = device.parameters.operationMode;
        return parameters;
    }
    getParameters(parameters) {
        const par = {};
        if (!parameters)
            return par;
        if (parameters.operate !== undefined)
            par.operate = parameters.operate;
        if (parameters.temperatureSet !== undefined)
            par.temperatureSet = parameters.temperatureSet;
        if (parameters.fanAutoMode !== undefined)
            par.fanAutoMode = parameters.fanAutoMode;
        if (parameters.airDirection !== undefined)
            par.airDirection = parameters.airDirection;
        if (parameters.airSwingLR !== undefined)
            par.airSwingLR = parameters.airSwingLR;
        if (parameters.airSwingUD !== undefined)
            par.airSwingUD = parameters.airSwingUD;
        if (parameters.fanSpeed !== undefined)
            par.fanSpeed = parameters.fanSpeed;
        if (parameters.actualNanoe !== undefined)
            par.actualNanoe = parameters.actualNanoe;
        if (parameters.nanoe !== undefined)
            par.nanoe = parameters.nanoe;
        if (parameters.ecoFunctionData !== undefined)
            par.ecoFunctionData = parameters.ecoFunctionData;
        if (parameters.ecoMode !== undefined)
            par.ecoMode = parameters.ecoMode;
        if (parameters.operationMode !== undefined)
            par.operationMode = parameters.operationMode;
        if (parameters.insideCleaning !== undefined)
            par.insideCleaning = parameters.insideCleaning;
        return par;
    }
    /**
     *
     * @param options RequestOprions to use
     * @param data optional data to use for request body
     * @returns Promise<any>
     */
    request(options, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const self = this;
            return yield new Promise((resolve, reject) => {
                const client = (options.protocol == "https:") ? https : http_1.default;
                try {
                    const req = client.request(options, (res) => {
                        let str = "";
                        res.on("data", (chunk) => {
                            str += chunk;
                        });
                        res.on("end", () => {
                            const response = this.JsonTryParse(str);
                            const statusCode = Number(res === null || res === void 0 ? void 0 : res.statusCode);
                            if (statusCode >= 200 && statusCode < 300) {
                                resolve(response);
                            }
                            else {
                                response.httpCode = res === null || res === void 0 ? void 0 : res.statusCode;
                                response.statusCode = res === null || res === void 0 ? void 0 : res.statusCode;
                                response.statusMessage = res === null || res === void 0 ? void 0 : res.statusMessage;
                                reject(response);
                            }
                        });
                    });
                    req.on("error", (e) => {
                        console.error(`problem with request: ${e.message}`);
                        reject(e);
                        req.destroy();
                    });
                    if (data) {
                        req.write(data);
                    }
                    req.end();
                }
                catch (error) {
                    console.log(`problem with request: ${error}`);
                    reject(error);
                }
            });
        });
    }
    /**
     *
     * @param method HTTP method to use
     * @param uri Uri to use
     * @returns An object containing request options
     */
    getRequestOptions(method, uri) {
        const requestOptions = {
            host: uri.host,
            port: uri.port,
            path: uri.path,
            protocol: uri.protocol,
            method: method,
            headers: {
                Connection: "Keep-Alive",
                "Content-Type": "application/json; charset=utf-8",
                Accept: "application/json; charset=utf-8",
                Host: uri.hostname,
                "X-APP-TYPE": 1,
                "X-APP-VERSION": this._ccAppVersion,
                "X-APP-NAME": "Comfort Cloud",
                "X-APP-TIMESTAMP": "0",
                "X-CFC-API-KEY": "Comfort Cloud",
                "User-Agent": "G-RAC",
            },
        };
        if (this._accessToken !== '42' && requestOptions.headers) {
            requestOptions.headers["X-User-Authorization"] = this._accessToken;
        }
        return requestOptions;
    }
    /**
     * Try to parse a string and return a valid JSON object.
     * If string is not valid JSON, it will return an empty object instead.
     * @param input Input string to try to parse as a JSON object
     * @returns Parsed or empty Json object
     */
    JsonTryParse(input) {
        try {
            //check if the string exists
            if (input) {
                let o = JSON.parse(input);
                //validate the result too
                if (o && o.constructor === Object) {
                    return o;
                }
            }
        }
        catch (e) {
        }
        return { responseMessage: input };
    }
}
exports.ComfortCloud = ComfortCloud;
