// panasonic-comfort-cloud-api/dist/models/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InsideCleaning = exports.DataMode = exports.NanoeMode = exports.FanAutoMode = exports.dataMode = exports.FanSpeed = exports.EcoMode = exports.AirSwingLR = exports.AirSwingUD = exports.OperationMode = exports.Power = void 0;
var Power;
(function (Power) {
    Power[Power["Off"] = 0] = "Off";
    Power[Power["On"] = 1] = "On";
})(Power || (exports.Power = Power = {}));
var OperationMode;
(function (OperationMode) {
    OperationMode[OperationMode["Auto"] = 0] = "Auto";
    OperationMode[OperationMode["Dry"] = 1] = "Dry";
    OperationMode[OperationMode["Cool"] = 2] = "Cool";
    OperationMode[OperationMode["Heat"] = 3] = "Heat";
    OperationMode[OperationMode["Fan"] = 4] = "Fan";
})(OperationMode || (exports.OperationMode = OperationMode = {}));
var AirSwingUD;
(function (AirSwingUD) {
    AirSwingUD[AirSwingUD["Up"] = 0] = "Up";
    AirSwingUD[AirSwingUD["UpMid"] = 3] = "UpMid";
    AirSwingUD[AirSwingUD["Mid"] = 2] = "Mid";
    AirSwingUD[AirSwingUD["DownMid"] = 4] = "DownMid";
    AirSwingUD[AirSwingUD["Down"] = 1] = "Down";
})(AirSwingUD || (exports.AirSwingUD = AirSwingUD = {}));
var AirSwingLR;
(function (AirSwingLR) {
    AirSwingLR[AirSwingLR["Left"] = 0] = "Left";
    AirSwingLR[AirSwingLR["LeftMid"] = 4] = "LeftMid";
    AirSwingLR[AirSwingLR["Mid"] = 2] = "Mid";
    AirSwingLR[AirSwingLR["RightMid"] = 3] = "RightMid";
    AirSwingLR[AirSwingLR["Right"] = 1] = "Right";
})(AirSwingLR || (exports.AirSwingLR = AirSwingLR = {}));
var EcoMode;
(function (EcoMode) {
    EcoMode[EcoMode["Auto"] = 0] = "Auto";
    EcoMode[EcoMode["Powerful"] = 1] = "Powerful";
    EcoMode[EcoMode["Quiet"] = 2] = "Quiet";
})(EcoMode || (exports.EcoMode = EcoMode = {}));
var FanSpeed;
(function (FanSpeed) {
    FanSpeed[FanSpeed["Auto"] = 0] = "Auto";
    FanSpeed[FanSpeed["Low"] = 1] = "Low";
    FanSpeed[FanSpeed["LowMid"] = 2] = "LowMid";
    FanSpeed[FanSpeed["Mid"] = 3] = "Mid";
    FanSpeed[FanSpeed["HighMid"] = 4] = "HighMid";
    FanSpeed[FanSpeed["High"] = 5] = "High";
})(FanSpeed || (exports.FanSpeed = FanSpeed = {}));
var dataMode;
(function (dataMode) {
    dataMode[dataMode["Day"] = 0] = "Day";
    dataMode[dataMode["Week"] = 1] = "Week";
    dataMode[dataMode["Month"] = 2] = "Month";
    dataMode[dataMode["Year"] = 4] = "Year";
})(dataMode || (exports.dataMode = dataMode = {}));
var FanAutoMode;
(function (FanAutoMode) {
    FanAutoMode[FanAutoMode["Disabled"] = 1] = "Disabled";
    FanAutoMode[FanAutoMode["AirSwingAuto"] = 0] = "AirSwingAuto";
    FanAutoMode[FanAutoMode["AirSwingLR"] = 3] = "AirSwingLR";
    FanAutoMode[FanAutoMode["AirSwingUD"] = 2] = "AirSwingUD";
})(FanAutoMode || (exports.FanAutoMode = FanAutoMode = {}));
var NanoeMode;
(function (NanoeMode) {
    NanoeMode[NanoeMode["Unavailable"] = 0] = "Unavailable";
    NanoeMode[NanoeMode["Off"] = 1] = "Off";
    NanoeMode[NanoeMode["On"] = 2] = "On";
    NanoeMode[NanoeMode["ModeG"] = 3] = "ModeG";
    NanoeMode[NanoeMode["All"] = 4] = "All";
})(NanoeMode || (exports.NanoeMode = NanoeMode = {}));
var DataMode;
(function (DataMode) {
    DataMode[DataMode["Day"] = 0] = "Day";
    DataMode[DataMode["Week"] = 1] = "Week";
    DataMode[DataMode["Month"] = 2] = "Month";
    DataMode[DataMode["Year"] = 4] = "Year";
})(DataMode || (exports.DataMode = DataMode = {}));
var InsideCleaning;
(function (InsideCleaning) {
    InsideCleaning[InsideCleaning["NotAvailable"] = 0] = "NotAvailable";
    InsideCleaning[InsideCleaning["Off"] = 1] = "Off";
    InsideCleaning[InsideCleaning["On"] = 2] = "On";
})(InsideCleaning || (exports.InsideCleaning = InsideCleaning = {}));
